sleepecg/io/sleep_readers.py

```diff
@@ -0,0 +1,269 @@
+    recording_start_time : datetime.time
+        Time at which the recording was started.
+
+    """
+    STAGE_MAPPING = {
+        'Wake|0': SleepStage.WAKE,
+        'Stage 1 sleep|1': SleepStage.N1,
+        'Stage 2 sleep|2': SleepStage.N2,
+        'Stage 3 sleep|3': SleepStage.N3,
+        'Stage 4 sleep|4': SleepStage.N3,
+        'REM sleep|5': SleepStage.REM,
+        'Unscored|9': SleepStage.UNDEFINED,
+    }
+
+    tree = ElementTree.parse(xml_filepath)
+    root = tree.getroot()
+
+    epoch_length = root.findtext('EpochLength')
+    if epoch_length is None:
+        raise RuntimeError(f'EpochLength not found in {xml_filepath}.')
+    epoch_length = int(epoch_length)
+
+    start_time = None
+    annot_stages = []
+
+    for event in root.find('ScoredEvents'):
+        if event.find('EventConcept').text == 'Recording Start Time':
+            start_time = event.find('ClockTime').text.split()[1]
+            start_time = datetime.datetime.strptime(start_time, '%H.%M.%S').time()
+
+        if event.find('EventType').text == 'Stages|Stages':
+            epoch_duration = int(float(event.findtext('Duration')))
+            stage = STAGE_MAPPING[event.findtext('EventConcept')]
+            annot_stages.extend([stage] * int(epoch_duration / epoch_length))
+
+    if start_time is None:
+        raise RuntimeError(f'"Recording Start Time" not found in {xml_filepath}.')
+
+    return _ParseNsrrXmlResult(
+        np.array(annot_stages, dtype=np.int8),
+        epoch_length,
+        start_time,
+    )
+
+
+def read_mesa(
+    data_dir: Optional[Union[str, Path]] = None,
+    records_pattern: str = '*',
+    use_preprocessed_heartbeats: bool = True,
+    offline: bool = False,
+    persist_edfs: bool = False,
+) -> Iterator[SleepRecord]:
+    """
+    Lazily reads records from MESA (https://sleepdata.org/datasets/mesa).
+
+    Each MESA record consists of a `.edf` files containing raw
+    polysomnography data and an `.xml` files containing annotated events.
+    Since the entire MESA dataset requires about 385 GB of disk space,
+    `.edf` files can be deleted after the heartbeat times have been
+    extracted. Heartbeat times are stored to a `.npy` file in
+    `<data_dir>/mesa/preprocessed/heartbeats`.
+
+    Parameters
+    ----------
+    data_dir : str | pathlib.Path
+        Directory where all datasets are stored. If `None` (default), the
+        value will be taken from the configuration.
+    records_pattern : str, optional
+         Glob-like pattern to select record IDs, by default `'*'`.
+    use_preprocessed_heartbeats : bool, optional
+        If `True`, get heartbeat times directly from the stored `.npy`
+        file, so `.edf` files are only downloaded for records which have
+        not yet been preprocessed. By default `True`.
+    offline : bool, optional
+        If `True`, search for local files only instead of using the NSRR
+        API, by default `False`.
+    persist_edfs : bool, optional
```

Comment:
Can we rename this parameter to `keep_edfs`? I think this is easier to understand.